github/transport.go

```diff
@@ -197,3 +197,99 @@ func isWriteMethod(method string) bool {
 	}
 	return false
 }
+
+type RetryTransport struct {
+	transport       http.RoundTripper
+	retryDelay      time.Duration
+	maxRetries      int
+	retryableErrors map[int]bool
+}
+
+type RetryTransportOption func(*RetryTransport)
+
+// NewRetryTransport takes in an http.RoundTripper and a variadic list of
+// optional functions that modify the RetryTransport struct itself. This
+// may be used to retry after response errors 5xx, for example.
+func NewRetryTransport(rt http.RoundTripper, options ...RetryTransportOption) *RetryTransport {
+	// Default to no retry if none is provided
+	defaultErrors := getDefaultRetriableErrors()
+	rlt := &RetryTransport{transport: rt, retryDelay: time.Second, maxRetries: 0, retryableErrors: defaultErrors}
+
+	for _, opt := range options {
+		opt(rlt)
+	}
+
+	return rlt
+}
+
+func (t *RetryTransport) RoundTrip(req *http.Request) (*http.Response, error) {
+	var err error
+	var resp *http.Response
+	var dataBuffer *bytes.Reader
+
+	for retry := 0; retry <= t.maxRetries; retry++ {
+		// Reset the body
+		// Code from httpretry (https://github.com/ybbus/httpretry/blob/master/roundtripper.go#L60)
```

Comment:
Can you talk to me a little bit about this code and why you chose it?

Reply:
How was this tested?

Reply:
The tests on `transport_test.go` were failing due to the body not being reset after the first retry, then I checked the package I use on some of my projects to retry (httpretry) and saw that they have a code for this scenario, so I shameless copied it and the tests succeed

I didn't figure out how to test the specific scenario of using the function `GetBody`, the original library also only tests without the [GetBody](https://github.com/ybbus/httpretry/blob/3adfad4cb89df5e87de20d1c0207c22c17b754df/roundtripper_test.go#L165), but if you have an idea on how to test with it I can update the tests

Reply:
Do you have a manual workload you can test this on that will reliably trigger a 500? I'm hesitant to merge/release to a wide audience without a bit more description of the validation process.

Reply:
I have a Flux integration that quite often gives me a 500 (a few times a day), I can try linking my branch instead of the default provider there... however, I have never integrated an in development provider, so I will appreciate anything that can guide me through this
I'm also a bit afraid of doing that on this workload, it is not critical, but I would expect there are more changes here that are not tested/stable

Reply:
You can create the binary for this provider and point Terraform towards it by seeing our [debugging instructions](https://github.com/integrations/terraform-provider-github/blob/main/CONTRIBUTING.md#debugging-the-terraform-provider). 

We do not have successful automated integration testing at the moment so we've been over-relying on manual testing. I don't feel super comfortable merging without manually testing this extensively first.